os/se/ameba/security_ameba_wrapper_tz.c

```diff
@@ -0,0 +1,826 @@
+
+	if (key_idx >= KEY_STORAGE_INDEX_MAX) {	/* Index Out of Range */
+		return HAL_INVALID_SLOT_RANGE;
+	}
+
+	in_output.input_data = hash;
+	in_output.output_data = sign;
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_ecdsa_sign_md(mode, &in_output, key_idx);
+	up_free_secure_context();
+
+	if (ret != HAL_SUCCESS) {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_ecdsa_verify_md(hal_ecdsa_mode mode, hal_data *hash, hal_data *sign, uint32_t key_idx)
+{
+	AWRAP_ENTER;
+	int ret;
+	inout_struc in_output;
+
+	if (key_idx >= KEY_STORAGE_INDEX_MAX) {	/* Index Out of Range */
+		return HAL_INVALID_SLOT_RANGE;
+	}
+
+	in_output.input_data = hash;
+	in_output.output_data = sign;
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_ecdsa_verify_md(mode, &in_output, key_idx);
+	up_free_secure_context();
+
+	if (ret != HAL_SUCCESS) {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_dh_generate_param(uint32_t dh_idx, hal_dh_data *dh_param)
+{
+	AWRAP_ENTER;
+	int ret;
+
+	if (dh_idx >= KEY_STORAGE_INDEX_MAX) {	/* Index Out of Range */
+		return HAL_INVALID_SLOT_RANGE;
+	}
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_dh_generate_param(dh_idx, dh_param);
+	up_free_secure_context();
+
+	if (ret != HAL_SUCCESS) {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_dh_compute_shared_secret(hal_dh_data *dh_param, uint32_t dh_idx, hal_data *shared_secret)
+{
+	AWRAP_ENTER;
+	int ret;
+
+	if (dh_idx >= KEY_STORAGE_INDEX_MAX) {	/* Index Out of Range */
+		return HAL_INVALID_SLOT_RANGE;
+	}
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_dh_compute_shared_secret(dh_param, dh_idx, shared_secret);
+	up_free_secure_context();
+
+	if (ret != HAL_SUCCESS) {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_ecdh_compute_shared_secret(hal_ecdh_data *ecdh_param, uint32_t key_idx, hal_data *shared_secret)
+{
+	AWRAP_ENTER;
+	int ret;
+
+	if (key_idx >= KEY_STORAGE_INDEX_MAX) {	/* Index Out of Range */
+		return HAL_INVALID_SLOT_RANGE;
+	}
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_ecdh_compute_shared_secret(ecdh_param, key_idx, shared_secret);
+	up_free_secure_context();
+
+	if (ret != HAL_SUCCESS) {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_set_certificate(uint32_t cert_idx, hal_data *cert_in)
+{
+	AWRAP_ENTER;
+	int ret;
+
+	if (cert_idx >= KEY_STORAGE_INDEX_MAX) {	/* Index Out of Range */
+		return HAL_INVALID_SLOT_RANGE;
+	}
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_set_certificate(cert_idx, cert_in);
+	up_free_secure_context();
+
+	if (ret != HAL_SUCCESS) {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_get_certificate(uint32_t cert_idx, hal_data *cert_out)
+{
+	AWRAP_ENTER;
+	int ret;
+
+	if (cert_idx == FACTORYKEY_RTL_CERT) {	/* If factory Cert */
+		ret = ameba_hal_get_factory_cert(cert_idx, cert_out);
```

Comment:
Is it okay not to call `up_allocate_secure_context()` and `up_free_secure_context()` before and after calling `ameba_hal_get_factory_cert()`?

Reply:
My mistake, added.